issuer-service/src/main/java/ch/admin/bj/swiyu/issuer/service/DemonstratingProofOfPossessionValidationService.java

```diff
@@ -0,0 +1,240 @@
+        }
+        // Note: ath will be not always be relevant
+        var mandatoryDpopPayloadClaims = List.of("jti", "htm", "htu", "iat", "nonce");
+        if (!jwtClaims.getClaims().keySet().containsAll(mandatoryDpopPayloadClaims)) {
+            throw new DemonstratingProofOfPossessionException("Missing mandatory JWT payload claims", DemonstratingProofOfPossessionError.INVALID_DPOP_PROOF);
+        }
+    }
+
+    private void hasMatchingHttpMethod(HttpRequest request, JWTClaimsSet jwtClaims) throws ParseException {
+        if (!StringUtils.equalsIgnoreCase(request.getMethod().name(), jwtClaims.getStringClaim("htm"))) {
+            throw new DemonstratingProofOfPossessionException("HTTP method mismatch between DPoP and request", DemonstratingProofOfPossessionError.INVALID_DPOP_PROOF);
+        }
+    }
+
+    private void hasValidSignature(SignedJWT dpopJwt, JWK key) throws JOSEException {
+        if (!dpopJwt.verify(new ECDSAVerifier(key.toECKey()))) {
+            throw new DemonstratingProofOfPossessionException("DPoP signature is invalid", DemonstratingProofOfPossessionError.INVALID_DPOP_PROOF);
+        }
+    }
+
+    private void hasDpopType(JWSHeader header) {
+        if (!DPOP_JWT_HEADER_TYP.equals(header.getType().toString())) {
+            throw new DemonstratingProofOfPossessionException("DPoP typ MUST be %s".formatted(DPOP_JWT_HEADER_TYP), DemonstratingProofOfPossessionError.INVALID_DPOP_PROOF);
+        }
+    }
+
+    private void matchesSupportedAlgorithms(JWSHeader header) {
+        var supportedAlgorithms = getSupportedAlgorithms();
+        if (!supportedAlgorithms.contains(header.getAlgorithm().getName())) {
+            throw new DemonstratingProofOfPossessionException("DPoP alg MUST be one of %s".formatted(
+                    String.join(",", supportedAlgorithms)),
+                    DemonstratingProofOfPossessionError.INVALID_DPOP_PROOF);
+        }
+    }
+
+    /**
+     * Check if the nonce is valid - not yet used and still within the acceptable time window
+     */
+    private void hasValidSelfContainedNonce(JWTClaimsSet jwtClaims) throws ParseException {
+        var nonce = new SelfContainedNonce(jwtClaims.getStringClaim("nonce"));
+        if (!nonce.isSelfContainedNonce() || !nonce.isValid(applicationProperties.getNonceLifetimeSeconds())) {
+            throw new DemonstratingProofOfPossessionException("Must use valid server provided nonce", DemonstratingProofOfPossessionError.INVALID_DPOP_PROOF);
+        }
+    }
+
+    private void hasMatchingHttpUri(HttpRequest request, JWTClaimsSet jwtClaims) throws URISyntaxException, ParseException {
+        if (isInvalidUrl(request.getURI(), jwtClaims.getStringClaim("htu"))) {
+            throw new DemonstratingProofOfPossessionException("URL mismatch between DPoP and request", DemonstratingProofOfPossessionError.INVALID_DPOP_PROOF);
+        }
+    }
+
+    private boolean isInvalidUrl(URI requestUri, String htu) throws URISyntaxException {
```

Comment:
```suggestion
    private boolean checkHtuClaim(URI requestUri, String htu) throws URISyntaxException {
```

I think this could be renamed to improve the readability